### PR TITLE
Add Unit Tests

### PR DESCRIPTION
**Pull request by kulcsarrudolf**
Hello @stefanalexlupu ,

I found your project on LinkedIn and I considered that this project can be a good chance to me to practice Jest. I hope this unit tests are helpful to you.
And besides that I did some modifications on **index.js:derange()** function.

P.S.: My formatting settings are different from yours. I recommend to set your settings in **package.js**. I already did some changes, but our settings are still different (see: README.MD and index.js). 

BR,
Rudolf

### CODE REVIEW

**Comment by stefanalexlupu** [comment]
Thank you @kulcsarrudolf for contributing! This project can really use some testing.

I will take the proper time to review these changes to make sure they are going in the direction I want this project to go.

File: package.json [hunk 1]
```diff
@@ -10,11 +10,24 @@
...
+  "keywords": [
+    "permutations",
+    "sets",
+    "derangement",
+    "math"
+  ],
   "author": "Lupu Stefan Alex",
-  "license": "ISC"
+  "license": "ISC",
+  "prettier": {
```
Comment (stefanalexlupu) [inline package.json]: If your focus here is adding tests, it would be great to just focus on that. Code style changes are so debatable and the single policy I want right now in this repo is "use standard ESLint". You can create a different PR for handling this, if you really want to, here are some recommendations:

* We can't really assume that everybody has prettier installed, we should also properly add it as dev dependency.
* Please add configuration in a dedicated file like `.prettierrc.json`
* Just to make sure that  there aren't any conflicts between ESLint configuration and Prettier we could use [something like this](https://github.com/prettier/eslint-config-prettier).

There are way too many changes related to code style so I'm just thinking that  if you want to focus on writing tests for this PR, you could just use the local ESLint config (as recommended). You said something about having differences; if you have some global config, please use the local config instead since if anybody else will get into this code it should be uniform.

File: README.md [hunk 2]
```diff
@@ -19,4 +19,4 @@ const derangedSet = derange([ /* your set here */ ])
 ### Note
 
 Since there is no possible derangement for empty sets and sets with a single
-element, the ouptut of the function for this cases will be the empty set.
\ No newline at end of file
+element, the ouptut of the function for this cases will be the empty set.
```
Comment (stefanalexlupu) [inline README.md]: Since you throw an exception now instead of returning the empty set, you could update this section as well.

File: src/index.js [hunk 3]
```diff
@@ -26,28 +26,25 @@ function fill (length) {
...
+    throw new Error('At least 2 elements are required')
   }
+
   const indices = shuffle(fill(length))
-  const displaced = displace(indices, 1)
-  const a = []
-  for (let i = 0; i < length; i++) {
-    a[indices[i]] = array[displaced[i]]
-  }
+  const derangedArray = displace(indices, 1)
```
Comment (stefanalexlupu) [inline src/index.js]: These edits break the function. It's my bad for not documenting the algorithm. In short, here's what is happening.

Since the only restriction that is imposed is that positions should no longer be the same, I perform a position shuffle instead of shuffling the entire array. This is supposed to save a ton of memory if the array is filled with objects instead of primitives. In this section I actually generate the new array based on the indices that I shuffled to fit the derangement criteria.

Here's the output of the function with this changes (using example.js):

```
a -> 2
b -> 0
c -> 5
d -> 1
e -> 6
f -> 4
g -> 3
```
It should instead be:

```
a -> c
b -> a
c -> f
d -> b
e -> g
f -> e
g -> d
```

Your tests pass because you compare `'a'` with `2` and expect it to not be true - which will always happen, even if you compare `'a'` with `0`.

Please revert changes in lines 46 -> 52. You can keep the name `derangedArray` for `a`, it's way better. Kudos to you on that one!

File: package.json [hunk 4]
```diff
@@ -10,11 +10,24 @@
     "eslint-plugin-import": "^2.22.0",
     "eslint-plugin-node": "^11.1.0",
     "eslint-plugin-promise": "^4.2.1",
-    "eslint-plugin-standard": "^4.0.1"
+    "eslint-plugin-standard": "^4.0.1",
+    "jest": "^26.2.2"
   },
   "scripts": {
+    "test": "jest"
```
Comment (stefanalexlupu) [inline package.json]: Please don't assume that everybody has jest installed globally and add it as a dev-dependency 😁

File: src/___test___/index.test.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
+const derange = require('./../index')
+
+test('test derangement of a set', () => {
+  const array = ['a', 'b', 'c', 'd', 'e', 'f', 'g']
+  const deranged = derange(array)
+
+  expect(deranged[0]).not.toBe('a')
```
Comment (stefanalexlupu) [inline src/___test___/index.test.js]: What about using something like this:

```
for (let i = 0; i < deranged.length; i++) {
    expect(deranged[i]).not.toBe(array[i])
  }
```

File: src/___test___/index.test.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  expect(deranged[1]).not.toBe('b')
+  expect(deranged[2]).not.toBe('c')
+  expect(deranged[3]).not.toBe('d')
+  expect(deranged[4]).not.toBe('e')
+  expect(deranged[5]).not.toBe('f')
+  expect(deranged[6]).not.toBe('g')
+})
+
+test('test deregement of a set with 2 elements', () => {
```
Comment (stefanalexlupu) [inline src/___test___/index.test.js]: This is redundant since you have the first test.

File: src/___test___/index.test.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  const array = ['a', 'b']
+  const deranged = derange(array)
+
+  expect(deranged[0]).not.toBe('a')
+  expect(deranged[1]).not.toBe('b')
+})
+
+test('Test when the array is empty', () => {
```
Comment (stefanalexlupu) [inline src/___test___/index.test.js]: It would be great if the test names were a bit more descriptive. Here something like:

> Empty set should throw exception

I searched for best practices when using jest but didn't find anything conclusive. If you use a pattern, please document it in a `README.md` file in the root test directory

File: src/___test___/index.test.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
+const derange = require('./../index')
+
```
Comment (stefanalexlupu) [inline src/___test___/index.test.js]: Since the result is different each time because of the random factor, there should also be a statistic test, like maybe generating 100 derangements and making sure that all pass the test.

File: src/___test___/index.test.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+test('Test when the array is empty', () => {
+  const array = []
+  expect(() => derange(array)).toThrow('At least 2 elements are required')
+})
+
+test('Test when the array contains only one element', () => {
+  const array = ['a']
+  expect(() => derange(array)).toThrow('At least 2 elements are required')
+})
```
Comment (stefanalexlupu) [inline src/___test___/index.test.js]: Why not simply use something like an `index.test.js` in the same folder as `index.js`, namely in the `src` folder?

**Review by stefanalexlupu — CHANGES_REQUESTED** [review]
Thank you for contributing, this PR is on the right path but some things need adjusting

* please remove code styling changes, since this PR only addresses the tests
* add the dependencies for modules used
* update documentation where necessary

**Comment by kulcsarrudolf** [comment]
@stefanalexlupu thank you for your feedback. I will make the changes based on your suggestions. Have a nice day!

File: package.json (hunk 1 above)
Comment (kulcsarrudolf) [reply package.json]: @stefanalexlupu do you use VS Code or something other? Could you please send me your code formatting settings?

File: src/index.js (hunk 3 above)
Comment (kulcsarrudolf) [reply src/index.js]: You're right. This means that my tests are also bad. I will improve this part.

File: package.json (hunk 1 above)
Comment (stefanalexlupu) [reply package.json]: I use VSCode, I think I should add an .editorconfig which might help, right?

File: src/index.js [hunk 10]
```diff
@@ -26,28 +26,25 @@ function fill (length) {
...
   return a
 }
 
-module.exports = function derange (array) {
+module.exports = function derange(array) {
   const length = array.length
 
   if (!length || length === 1) {
-    return []
+    throw new Error('At least 2 elements are required')
```
Comment (stefanalexlupu) [inline src/index.js]: I've just had a look at this kind of change. It's actually a breaking change and if we make it we should do it separately in a v2.* so I'll just create an issue for it to treat it in a next release.

You can [read more about this here](https://docs.npmjs.com/about-semantic-versioning)

I don't want to pretend like this is a huge library because it is not haha 😂 but if there's a code I think we should follow it. Especially since there's now almost 100 downloads of this package 🎉

File: package.json (hunk 1 above)
Comment (kulcsarrudolf) [reply package.json]: I don't have much experience with VS Code. I looked a little further, and I think that an **.editorconfig** can be usefull. It is important to define an uniform code format for this project.

**Review by stefanalexlupu — CHANGES_REQUESTED** [review]

**Comment by stefanalexlupu** [comment]
@kulcsarrudolf will you continue work on this PR?

**Comment by kulcsarrudolf** [comment]
> @kulcsarrudolf will you continue work on this PR?

Sorry, I was busy and I forgot to work on this PR. I will update it today.

File: src/index.js (hunk 3 above)
Comment (kulcsarrudolf) [reply src/index.js]: I reverted my changes.

File: src/___test___/index.test.js (hunk 6 above)
Comment (kulcsarrudolf) [reply src/___test___/index.test.js]: I think it is nice to have tests with edge-cases.

File: src/___test___/index.test.js (hunk 8 above)
Comment (kulcsarrudolf) [reply src/___test___/index.test.js]: Good idea! :)

File: src/___test___/index.test.js (hunk 9 above)
Comment (kulcsarrudolf) [reply src/___test___/index.test.js]: Maybe in the future you will have more test files for **index.js**.

File: src/index.js (hunk 10 above)
Comment (kulcsarrudolf) [reply src/index.js]: Ok. 
100 downloads 😮😎 , that is very nice! Congratulations! 👏👏 Keep it up!

**Comment by kulcsarrudolf** [comment]
@stefanalexlupu I did some modification, but I want to check it again today.

I think I found a little bug, it is ok if I create an Issue?  

Unfortunately I squashed my commits, I just realized it wasn't a good idea. 😞

**Comment by stefanalexlupu** [comment]
@kulcsarrudolf yes please create the issue :)

**Comment by kulcsarrudolf** [comment]
@stefanalexlupu please review my last modifications.

Some tests are skipped, because the tested cases are not implemented. But I recomand you to cosider these tests for future versions.

image

File: src/__test__/index.statistic.test.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+      'Watson',
+      'Noah',
+      'Ensley',
+      'Olivia',
+      'Benjamin',
+      'Evelyn'
+    ]
+  ]
+
+  cases.forEach((element) => {
```
Comment (stefanalexlupu) [inline src/__test__/index.statistic.test.js]: Element is such a generic name for a variable that actually contains an array... please change it to `case` or `set`.

File: src/__test__/index.statistic.test.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    for (let i = 0; i < 100; i++) {
+      const setToDerange = element
```
Comment (stefanalexlupu) [inline src/__test__/index.statistic.test.js]: Since you will never be mutating that `setToDerange` you don't really need to create this constant. Just stick with the value passed in the `forEach` callback to avoid clutter.

File: src/__test__/index.statistic.test.js [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+      let notDerangedElements = setToDerange
+      const derangedElements = []
+      const derangedSet = derange(setToDerange)
+
+      for (let i = 0; i < setToDerange.length; i++) {
+        expect(derangedSet[i]).not.toBe(setToDerange[i])
+        expect(setToDerange).toEqual(expect.arrayContaining([derangedSet[i]]))
```
Comment (stefanalexlupu) [inline src/__test__/index.statistic.test.js]: According to docs: https://jestjs.io/docs/en/expect#expectarraycontainingarray

With this check you could miss the case in which not all the initial elements made it to the derangedSet.

Why not do something like:

```
expect(setToDerange).toEqual(expect.arrayContaining(derangedSet))
expect(derangedSet).toEqual(expect.arrayContaining(setToDerange))
```

outside the `for` loop

This way you make sure that all elements that went into the derange method make it out (just in a different order).

image

File: src/__test__/index.statistic.test.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
```
Comment (stefanalexlupu) [inline src/__test__/index.statistic.test.js]: If I understand correctly, from this point onwards you are doing the following:

1. Add each element that passes the test above into the `derangedElements` array
2. Set the `notDerangedElements` to be the rest of the `derangedSet` (do you see the irony here?)
3. Check that the derangedElements is not a subset of the set containing only the current element?  When is this ever going to fail?

I'm waiting for you to explain to me what this is looking for but it seems to me that just performing the check at line 28 + the suggested test above is enough.

If that's the case in the end, we might not need

* `derangedElements`
* `notDerangedElements`
* `removeSpecificElementFromArray`
* `expect(notDerangedElements.length).toBe(EMPTY)` (or that `EMPTY` constant for that matter)

File: src/__test__/index.test.js [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+  for (let i = 1; i < testArray.length; i++) {
+    expect(derangedElements).not.toEqual(
+      expect.arrayContaining([derangement[i]])
+    )
+
+    derangedElements.push(derangement[i])
+  }
+})
+
+test('Test if all of the given array elements are deranged', () => {
```
Comment (stefanalexlupu) [inline src/__test__/index.test.js]: This might be simpler to read and understand

```
expect(setToDerange).toEqual(expect.arrayContaining(derangedSet))
expect(derangedSet).toEqual(expect.arrayContaining(setToDerange))
```

File: src/__test__/index.test.js [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+
+  expect(derangement[0]).not.toBe(arrayWithTwoElements[0])
+  expect(derangement[1]).not.toBe(arrayWithTwoElements[1])
+})
+
+test.skip('When the given array is empty should throw an error', () => {
+  const array = []
+  const errorMessage = 'At least 2 elements are required'
+
+  expect(() => derange(array)).toThrowError(errorMessage)
```
Comment (stefanalexlupu) [inline src/__test__/index.test.js]: We dropped this feature for now at least..  I logged an issue ( #4 ) for this and will add the test together with that.

It's going to be a breaking change so it will probably only happen later.

File: src/__test__/utils/utils.js [hunk 17]
```diff
@@ -0,0 +1,16 @@
+const utils = {
```
Comment (stefanalexlupu) [inline src/__test__/utils/utils.js]: The need of test utils scares the fire out of me :))

We have to make sure we need this as per above comments. I have a strong feeling we can live without these, although I really appreciate the effort!

File: src/index.js [hunk 18]
```diff
@@ -29,7 +29,7 @@ function fill (length) {
 function displace (array, by) {
   const length = array.length
   const a = []
-  array.forEach(i => {
+  array.forEach((i) => {
```
Comment (stefanalexlupu) [inline src/index.js:32]: Thank you for doing this 🙌🏻

**Review by stefanalexlupu — CHANGES_REQUESTED** [review]
I really appreciate your effort but it still seems like these tests can be improved.

I asked some questions so let's discuss and see if I understood everything correctly :)

File: src/__test__/index.statistic.test.js (hunk 11 above)
Comment (kulcsarrudolf) [reply src/__test__/index.statistic.test.js]: I deleted the **index.statistic.test.js** file.

File: src/__test__/utils/utils.js (hunk 17 above)
Comment (kulcsarrudolf) [reply src/__test__/utils/utils.js]: Yes, this is very strange. I deleted it :satisfied:

6b4upqv6at321

File: src/index.js (hunk 18 above)
Comment (kulcsarrudolf) [reply src/index.js:32]: This is your merit. The linter did it automatically. I like it! The pre-commit task is very useful.

File: src/__test__/index.test.js (hunk 16 above)
Comment (kulcsarrudolf) [reply src/__test__/index.test.js]: TDD :sunglasses:
I think it is better to keep it, but it's your decision.

**Comment by kulcsarrudolf** [comment]
@stefanalexlupu thank you for your constructive comments and for your time! 😃🙏

File: src/__test__/index.test.js (hunk 16 above)
Comment (stefanalexlupu) [reply src/__test__/index.test.js]: I will create a branch for v2 after this is done and will notify you to implement this if you want 😁

**Review by stefanalexlupu — COMMENTED** [review]
image

I tried running the test on my local machine and it seems to not work, do you know what this error is?

I only cloned your branch and run:

```
npm i
npm run test
```

**Comment by stefanalexlupu** [comment]
Update: it seems to be working with jest ^25, looking to see what changed

---

Edit: I looked at [the release docs for v26](https://jestjs.io/blog/2020/05/05/jest-26) and it seems to me that they are in the process of changing how methods are loaded into context when executing jest. I'm assuming you tested using a local installation of jest? Can you see what that is?

Maybe lock the version t v25 for now to avoid these changes?

**Review by stefanalexlupu — APPROVED** [review]
Thank you for your contribution @kulcsarrudolf!
